src/pydicom/pixels/utils.py

```diff
@@ -2098,3 +2170,246 @@ def unpack_bits(src: bytes, as_array: bool = True) -> "np.ndarray | bytes":
+    if pad and len(out) % 2 == 1:
+        out = b"".join([out, b"\x00"])
+
+    return out
+
+
+def concatenate_packed_frames(
+    frames: Sequence[bytes | bytearray],
+    frame_length: int,
+    pad: bool = True,
+) -> bytes:
+    """Concatenate individual bit-packed frames.
+
+    This should only be used for single bit data (Bits Allocated = 1) in a
+    "bit-packed" representation (one byte containing 8 pixels). Concatenating
+    such frames may require combining pixels from multiple frames within a
+    single byte.
+
+    Parameters
+    ----------
+    frames : Sequence[bytes | bytearray]
+        List of individual frames in native bit-packed representation.
+    frame_length : int
+        Number of pixels in each frame.
+    pad : bool
+        Whether to zero-pad the resulting bytes to a multiple of 2 bytes.
+
+    Returns
+    -------
+    bytes
+        Bit-packed representation of all frames combined.
+
+    Note
+    ----
+    In certain situations, this function will run faster and use less memory if
+    NumPy is installed. Otherwise, it will fall back to a slower method using
+    native Python functions.
+    """
+    if frame_length % 8 == 0:
+        # Simple concatenation since frame boundaries occur on byte boundaries
+        out = b"".join(frames)
+    else:
+        if HAVE_NP:
+            # Efficient implementation using NumPy to shift bits between bytes
+
+            # Pre-allocate an output array
+            total_bytes = math.ceil(len(frames) * frame_length / 8)
+            bytes_per_frame = math.ceil(frame_length / 8)
+            if pad and total_bytes % 2 == 1:
+                total_bytes += 1
+            out_arr = np.zeros(total_bytes, dtype=np.uint8)
+
+            # Temporary buffer re-used in the loop
+            tmp = np.empty(bytes_per_frame, dtype=np.uint8)
+
+            for i, frame in enumerate(frames):
+                pixel_start = i * frame_length
+                bit_offset = pixel_start % 8
+
+                # Index of the byte containing the first pixel of this frame
+                b = math.floor(pixel_start / 8)
+
+                # Input frames may be padded. Reading the whole buffer may take
+                # us beyond the end of the output array later, so only read the
+                # necessary bytes
+                frame_array = np.frombuffer(
+                    frame,
+                    count=bytes_per_frame,
+                    dtype=np.uint8,
+                )
+
+                if bit_offset != 0:
+                    # Need to shift bits between bytes
+
+                    # Combine most significant bytes of this frame's first byte
+                    # with least significant bits of the previous frame's last
+                    # byte
+                    np.bitwise_or(
+                        out_arr[b : b + 1],
+                        np.left_shift(frame_array[:1], bit_offset),
+                        out=out_arr[b : b + 1],
+                    )
```

Comment:
`out_arr[b] ^= (frame_array[0] << bit_offset)` is a bit easier to understand

Reply:
Addressed in [e9ea3de](https://github.com/pydicom/pydicom/pull/2216/commits/e9ea3deb538f6786f5a80e0e826d7aaf7a40678c)